<h2>Why is this an issue?</h2>

<p>This rule verifies that interface names follow the convention of ending with the 'Interface' suffix.</p>

<h3>Why is this important?</h3>
<p>Following naming conventions for interfaces provides several benefits:</p>
<ul>
    <li>Improves code readability by clearly identifying the purpose of the declaration</li>
    <li>Makes it easier to distinguish interfaces from classes at a glance</li>
    <li>Follows object-oriented best practices and established conventions</li>
    <li>Maintains consistency across the codebase</li>
</ul>

<h3>Noncompliant code example</h3>
<pre>
interface Customer {
    public function getName(): string;
    public function getEmail(): string;
}
</pre>

<h3>Compliant example</h3>
<pre data-diff-id="1" data-diff-type="compliant">
interface CustomerInterface {
    public function getName(): string;
    public function getEmail(): string;
}
</pre>

<h3>Exceptions</h3>
<p>There are no exceptions to this rule. All interfaces should follow this naming convention.</p>