<h2>Why is this an issue?</h2>
<p>
    In Magento 2, using hardcoded (literal) namespace strings in PHP code can reduce flexibility and maintainability.
    Instead, dynamic approaches such ::class should be used to improve code
    robustness, prevent typos, and facilitate refactoring.
</p>

<h3>Noncompliant code example</h3>
<pre data-diff-id="2" data-diff-type="noncompliant">
    public function notOk()
    {
        $this->create('Magento\CustomerSegment\Model\Segment\Condition\Customer\Address'); // Noncompliant
    }
</pre>

<h3>Compliant example</h3>
<pre data-diff-id="1" data-diff-type="compliant">
  public function ok()
    {
        $this->create(Magento\CustomerSegment\Model\Segment\Condition\Customer\Address::class); // Compliant
    }
</pre>