<h2>Why is this an issue?</h2>
<p>
    Templates MUST NOT instantiate objects. All objects MUST be passed from the Block objects.
</p>

<h3>Noncompliant code example</h3>
<pre data-diff-id="2" data-diff-type="noncompliant">
$customer = new \Magento\Customer\Model\Customer(); // Noncompliant
</pre>

<h3>Compliant solution</h3>
<pre data-diff-id="1" data-diff-type="compliant">
$customer = $this->getCustomer(); // Compliant
</pre>